### ISSUE TITLE
Use ByteVectorChunk?

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#876]
`ByteVectorChunk` is a type of `Chunk[Byte]` which is better optimized for avoiding copies.  Based on our experience with byte copying in scalaz-stream, I have a feeling this will be a big win for us.

But it only exists in fs2-1.0-SNAPSHOT.  We should think about how we want to cut over, or whether we want to kindly ask for an fs2-0.10 so we can start building on that.

**Comment by ChristopherDavenport** [comment]
Do we have an approximate eta on 1.0? This is a pretty big shift in how the implementation would work.

**Comment by rossabaker** [comment]
Yes, this is going to hit all the blaze projects hard in particular. I think we should ask the fs2 folks for advice.

**Comment by ChristopherDavenport** [comment]
FYI in gitter I got an approximate release date which is later than when we were hoping for. so I think we should hold off and look at later optimization, especially considering our currently slower benchmarks, unless fs2 could be pursuaded into a 0.10.

**Comment by ChristopherDavenport** [comment]
Is this in the 0.10 build that is being worked on in #1197 since it makes life much easier in a lot of places.

**Comment by aeons** [comment]
`ByteVectorChunk` is in `fs2-scodec`. I haven't changed any code that wasn't using it to begin with though.

The `topic/fast` stuff in fs2 moves the focus to using segments instead of chunks, where applicable, so this might lead to more restructuring than just `ByteVectorChunk`.

**Comment by rossabaker** [comment]
If we can avoid copying bytes with the segment code (I don't have any experience here), we might find that `ByteVectorChunk` no longer carries its weight.

**Comment by jmcardon** [comment]
Is this still relevant after #1588 ?

I don't think we do much `toArray` anymore, if at all.

**Comment by rossabaker** [comment]
I would love to explore dropping that dependency with the advances in fs2

**Comment by aeons** [comment]
Some benchmarks would be great for these decisions.

**Comment by rossabaker** [comment]
Here's a crude benchmark for concatenating some chunks and rendering it to an array for writing.

```scala
package org.http4s.bench

import fs2._
import fs2.interop.scodec.ByteVectorChunk
import org.openjdk.jmh.annotations._
import scodec.bits.ByteVector

@Fork(2)
@Threads(1)
@State(Scope.Benchmark)
@BenchmarkMode(Array(Mode.Throughput))
class BytesBench {
  val bs: Array[Byte] = "The quick brown fox jumped over the lazy dog.".getBytes("UTF-8")

  @Benchmark def byteVectorChunk = {
    val s = Segment.chunk(ByteVectorChunk(ByteVector.view(bs))) ++
      Segment.chunk(ByteVectorChunk(ByteVector.view(bs))) ++
      Segment.chunk(ByteVectorChunk(ByteVector.view(bs))) ++
      Segment.chunk(ByteVectorChunk(ByteVector.view(bs))) ++
      Segment.chunk(ByteVectorChunk(ByteVector.view(bs)))
    s.force.toArray
  }

  @Benchmark def bytes = {
    val s = Chunk.bytes(bs).toSegment ++
      Chunk.bytes(bs).toSegment ++
      Chunk.bytes(bs).toSegment ++
      Chunk.bytes(bs).toSegment ++
      Chunk.bytes(bs).toSegment
    s.force.toArray
  }
}
```

```
[info] Benchmark                    Mode  Cnt       Score       Error  Units
[info] BytesBench.byteVectorChunk  thrpt   40  350288.096 ± 12782.808  ops/s
[info] BytesBench.bytes            thrpt   40  470374.904 ± 19307.606  ops/s
```

**Comment by rossabaker** [comment]
I have created a `ByteBufferChunk` that is a `Chunk[Byte]` backed directly by a `ByteBuffer`.  Wrapping `ByteBuffer` is the primary use of `ByteVectorChunk`.  I replaced everything except its use in multipart, where fast concatenation and splitting are key to the current implemenation.

We can eliminate the fs2-scodec and scodec-bits dependencies if we redo multipart.  Less ambitiously, we could inhale a copy of `ByteVector` until we redo multipart, and shed he dependency for 0.18.0.  Thoughts?

**Comment by rossabaker** [comment]
functional-streams-for-scala/fs2#1032 should perform well for us, eliminate the chunk ops, and drop right into most uses other than multipart and some tests.

**Comment by jmcardon** [comment]
I think this is possibly closed by #1701 as well.

**Comment by rossabaker** [comment]
Yes, it will.
